subprojects/core-api/src/main/java/org/gradle/api/file/ConfigurableFileCollection.java

```diff
@@ -81,4 +84,43 @@ public interface ConfigurableFileCollection extends FileCollection, HasConfigura
      * @return this
      */
     ConfigurableFileCollection builtBy(Object... tasks);
+
+    /**
+     * Applies an eager transformation to the current contents of this file collection, without explicitly resolving it.
+     * The provided transformer is applied to the file collection representing the current contents, and the returned collection is used as a new content.
+     * The current contents collection can be used to derive the new value, but doesn't have to.
+     * Returning null from the transformer empties this collection.
+     * For example, it is possible to filter out all text files from the collection:
+     * <pre class='autoTested'>
+     *     def collection = files("a.txt", "b.md")
+     *
+     *     collection.update { it.filter { f -&gt; !f.name.endsWith(".txt") } }
+     *
+     *     println(collection.files) // ["b.md"]
+     * </pre>
+     * <p>
+     * <b>Further changes to this file collection, such as calls to {@link #setFrom(Object...)} or {@link #from(Object...)}, are not transformed, and override the update instead</b>.
+     * Because of this, this method inherently depends on the order of changes, and therefore must be used sparingly.
+     * <p>
+     * If this file collection consists of other mutable sources, then the current contents collection tracks the changes to these sources.
+     * For example, changes to the upstream collection are visible:
+     * <pre class='autoTested'>
+     *     def upstream = files("a.txt", "b.md")
+     *     def collection = files(upstream)
+     *
+     *     collection.update { it.filter { f -&gt; !f.name.endsWith(".txt") } }
+     *     upstream.from("c.md", "d.txt")
+     *
+     *     println(collection.files) // ["b.md", "c.md"]
+     * </pre>
+     * The provided transformation runs <b>eagerly</b>, so it can capture any objects without introducing memory leaks and breaking configuration cache.
+     * However, transformations applied to the current contents collection (like {@link FileCollection#filter(Closure)}) are subject to the usual constraints.
+     * <p>
+     * The current contents collection inherits dependencies of this collection specified by {@link #builtBy(Object...)}.
+     *
+     * @param transform the transformation to apply to the current value. May return null, which empties this collection.
+     * @since 8.6
+     */
+    @Incubating
+    void update(Transformer<? extends @org.jetbrains.annotations.Nullable FileCollection, ? super FileCollection> transform);
```

Comment:
@bamboo Not feedback for this PR per se, but this could end up being pulled up into the `Updatable` interface that will be introduced later.

 `void update(Transformer<? extends @Nullable Updatable, ? super Updatable> transform);`
 
 so this contract can be at one place (then further refined with type-specific semantics where needed).

Reply:
It would be good to have this definition with its sizable javadoc in a single place.